hypervisor/src/mshv/mod.rs

```diff
@@ -405,6 +405,8 @@ pub struct MshvVcpu {
     msrs: Vec<MsrEntry>,
     vm_ops: Option<Arc<dyn vm::VmOps>>,
     vm_fd: Arc<VmFd>,
+    #[cfg(feature = "sev_snp")]
+    ghcb_addr: Option<u64>,
```

Comment:
It looks like we `.unwrap()` and cast this to `*mut svm_ghcb_base` everywhere it is used.
If it's not optional, and it's never used as a `u64`, then can't we just make this a `*mut svm_ghcb_base` instead?

Reply:
It's optional, cause we might need to build the binary to run both CVM and non CVM. In non-cvm case we do not want to mmap the GHCB page but sev-snp feature flag is still on.

Reply:
Gotcha, ok. But maybe it can be a `Option<*mut svm_ghcb_base>`

Reply:
Not sure how to use it. I get compilation error:
*mut mshv_bindings::svm_ghcb_base` cannot be sent between threads safely

Reply:
You can wrap that option inside `Arc`.

Reply:
> You can wrap that option inside `Arc`.

I think we'd want the `Arc` inside the `Option`, so you don't need to access the Arc just to check if something is there

Reply:
Raw pointer is `!Send`. That's why the compiler is not happy. Using a `u64` works around the issue. Using `Arc` is likely overkilled.

You can do `struct Ghcb(*mut svm_ghcb_base); unsafe impl Send for Ghcb;`, and then use `Option<Ghcb>`. You do need to clearly reason why it is okay to implement `Send`.

With a new struct, you can also implement various methods and confine all the `unsafe`s into a few locations.

Reply:
> `struct Ghcb(*mut svm_ghcb_base); unsafe impl Send for Ghcb;`

+1
Seems like a better option than `u64` or raw pointer in Arc...

Reply:
Since we are using ghcb address per VCPU, I don't see any need to use Arc here.

Reply:
This needs MSHV changes and release to mshv and vfio as well.

Reply:
We can put the code in Cloud Hypervisor.

Reply:
I just saw your PR to the MSHV crate. We have a structure there. I'm fine with doing that over there, too.

Reply:
Compiler is asking to implement send/sync for the inner struct:

1287 |         self.set_fpu(&state.fpu)?;
     |              ^^^^^^^ `*mut mshv_bindings::svm_ghcb_base` cannot be sent between threads safely
     |
     = help: within `mshv::MshvVcpu`, the trait `std::marker::Send` is not implemented for `*mut mshv_bindings::svm_ghcb_base`, which is required by `mshv::MshvVcpu: std::marker::Send`
     = help: the trait `std::marker::Send` is implemented for `mshv_bindings::svm_ghcb_base`
note: required because it appears within the type `mshv::Ghcb`